osf/migrations/0197_draft_node_permissions.py

```diff
@@ -0,0 +1,76 @@
+      );
+
+    DELETE FROM auth_group WHERE name in
+    (SELECT regexp_split_to_table('draft_registration_' || D.id || '_read,draft_registration_' || D.id || '_write,draft_registration_' || D.id || '_admin', ',')
+    FROM osf_draftregistration D);
+    """
+
+# Forward migration - add read permissions to all draft registration django read groups, add read/write perms
+# to all draft registration django write groups, and add read/write/admin perms to all draft registration django admin groups
+add_permissions_to_draft_registration_groups = """
+    -- Adds "read_draft_registration" permissions to all Draft Reg read groups - uses DraftRegistrationGroupObjectPermission table
+    INSERT INTO osf_draftregistrationgroupobjectpermission (content_object_id, group_id, permission_id)
+    SELECT D.id as content_object_id, G.id as group_id, PERM.id AS permission_id
+    FROM osf_draftregistration AS D, auth_group G, auth_permission AS PERM
+    WHERE G.name = 'draft_registration_' || D.id || '_read'
+    AND PERM.codename = 'read_draft_registration';
+
+    -- Adds "read_draft_registration" and "write_draft_registration" permissions to all Draft Reg write groups
+    INSERT INTO osf_draftregistrationgroupobjectpermission (content_object_id, group_id, permission_id)
+    SELECT D.id as content_object_id, G.id as group_id, PERM.id AS permission_id
+    FROM osf_draftregistration AS D, auth_group G, auth_permission AS PERM
+    WHERE G.name = 'draft_registration_' || D.id || '_write'
+    AND (PERM.codename = 'read_draft_registration' OR PERM.codename = 'write_draft_registration');
+
+    -- Adds "read_draft_registration", "write_draft_registration", and "admin_draft_registration" permissions to all Draft Reg admin groups
+    INSERT INTO osf_draftregistrationgroupobjectpermission (content_object_id, group_id, permission_id)
+    SELECT D.id as content_object_id, G.id as group_id, PERM.id AS permission_id
+    FROM osf_draftregistration AS D, auth_group G, auth_permission AS PERM
+    WHERE G.name = 'draft_registration_' || D.id || '_admin'
+    AND (PERM.codename = 'read_draft_registration' OR PERM.codename = 'write_draft_registration' OR PERM.codename = 'admin_draft_registration');
+    """
+
+# Reverse migration - Remove all rows from DraftRegistrationGroupObjectPermission table - table gives draft django groups
+# permissions to draft reg
+drop_draft_reg_group_object_permission_table = """
+    DELETE FROM osf_draftregistrationgroupobjectpermission;
+    """
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('osf', '0196_draft_node_models'),
+    ]
+
+    operations = [
+        migrations.RunPython(post_migrate_signal, migrations.RunPython.noop),
+        migrations.RunSQL(add_draft_read_write_admin_auth_groups, remove_draft_auth_groups),
+        migrations.RunSQL(add_permissions_to_draft_registration_groups, drop_draft_reg_group_object_permission_table),
```

Comment:
I'm not sure - Is there a good way for me to estimate the time it will take or is it something someone with access to the prod database will have to test?

Reply:
It's mostly going to depend on how many rows there will be from the draftregistrations table being used to construct these queries. It will be worth running on prod data before we go, which will probably involve me getting some updated prod data.  My guess is that there aren't all that many active draft registrations at any given time, but we'll see.

I don't think we've merged that new flag for data migrations, but we should have chosen what it was by now. When we get the next B&I release out, we'll definitely have that flag. So hopefully sometime this month.

Reply:
looks like the SQL above acts on *all* draft registrations, not just active ones. worth running on prod data to benchmark, but making the data migration conditional can be addressed in a later PR.